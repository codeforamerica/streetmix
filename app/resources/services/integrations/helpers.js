const { User } = require('../../../db/models')

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
const findUser = async function (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
const addOrUpdateByProviderName = function (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
const syncAccountStatus = async function (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

module.exports = { findUser, addOrUpdateByProviderName, syncAccountStatus }
