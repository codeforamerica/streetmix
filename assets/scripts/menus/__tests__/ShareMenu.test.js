/* eslint-env jest */
import React from 'react'
import { cleanup } from '@testing-library/react'
import { renderWithReduxAndIntl } from '../../../../test/helpers/render'
import ShareMenu from '../ShareMenu'

describe('ShareMenu', () => {
  afterEach(cleanup)

  it('renders (user not signed in)', () => {
    const wrapper = renderWithReduxAndIntl(<ShareMenu />, { initialState: { user: { signedIn: false } } })
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders (user signed in)', () => {
    const wrapper = renderWithReduxAndIntl(<ShareMenu />, { initialState: { user: { signedIn: true } } })
    expect(wrapper.asFragment()).toMatchSnapshot()
  })
})
