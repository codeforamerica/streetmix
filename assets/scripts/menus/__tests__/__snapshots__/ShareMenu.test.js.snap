// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareMenu renders (user not signed in) 1`] = `
<DocumentFragment>
  <div
    class="menu share-menu"
  >
    <div
      class="share-sign-in-promo"
    >
      <span>
        <a
          href="#"
        >
          <span>
            Sign in
          </span>
        </a>
         for nicer links to your streets and your personal street gallery
      </span>
    </div>
    <div
      class="share-via-link-container"
    >
      <span>
        Copy and paste this link to share:
      </span>
      <input
        class="share-via-link"
        readonly=""
        spellcheck="false"
        type="text"
        value="http://localhost/"
      />
    </div>
    <a
      class="share-via-twitter"
      href="https://twitter.com/intent/tweet?text=Check%20out%20this%20street%20on%20Streetmix!&url=http%3A%2F%2Flocalhost%2F"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        class="svg-inline--fa fa-twitter"
      />
      <span>
        Share using Twitter
      </span>
    </a>
    <a
      class="share-via-facebook"
      href="https://www.facebook.com/dialog/feed?app_id=undefined&redirect_uri=http%3A%2F%2Flocalhost%2F&link=http%3A%2F%2Flocalhost%2F&name=Unnamed%20St%20%E2%80%93%20Streetmix&description=Check%20out%20this%20street%20on%20Streetmix!"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        class="svg-inline--fa fa-facebook-square"
      />
      <span>
        Share using Facebook
      </span>
    </a>
    <a
      href="#"
    >
      <span>
        Print…
      </span>
    </a>
    <a
      href="#"
      id="save-as-image"
    >
      <span>
        Save as image…
      </span>
      <span
        class="menu-item-subtext"
      >
        <span>
          For including in a report, blog, etc.
        </span>
      </span>
    </a>
  </div>
</DocumentFragment>
`;

exports[`ShareMenu renders (user signed in) 1`] = `
<DocumentFragment>
  <div
    class="menu share-menu"
  >
    <div
      class="share-via-link-container"
    >
      <span>
        Copy and paste this link to share:
      </span>
      <input
        class="share-via-link"
        readonly=""
        spellcheck="false"
        type="text"
        value="http://localhost/"
      />
    </div>
    <a
      class="share-via-twitter"
      href="https://twitter.com/intent/tweet?text=Check%20out%20this%20street%20on%20Streetmix!&url=http%3A%2F%2Flocalhost%2F"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        class="svg-inline--fa fa-twitter"
      />
      <span>
        Share using Twitter
      </span>
    </a>
    <a
      class="share-via-facebook"
      href="https://www.facebook.com/dialog/feed?app_id=undefined&redirect_uri=http%3A%2F%2Flocalhost%2F&link=http%3A%2F%2Flocalhost%2F&name=Unnamed%20St%20%E2%80%93%20Streetmix&description=Check%20out%20this%20street%20on%20Streetmix!"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        class="svg-inline--fa fa-facebook-square"
      />
      <span>
        Share using Facebook
      </span>
    </a>
    <a
      href="#"
    >
      <span>
        Print…
      </span>
    </a>
    <a
      href="#"
      id="save-as-image"
    >
      <span>
        Save as image…
      </span>
      <span
        class="menu-item-subtext"
      >
        <span>
          For including in a report, blog, etc.
        </span>
      </span>
    </a>
  </div>
</DocumentFragment>
`;
